File: p001-p099/leetcode004.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
*                     ----Stay Hungry Stay Foolish----                       *
*   @author	:   Shen                                                         *
*   @name   :   Leetcode 004 Median of Two Sorted Arrays                     *
*****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template<class T>inline bool updateMin(T& a, T b){ return a > b ? a = b, 1: 0; }
template<class T>inline bool updateMax(T& a, T b){ return a < b ? a = b, 1: 0; }

class Solution {
public:
	double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2)
	{
		int l1 = nums1.size(), l2 = nums2.size();
		vector<int> &sh = l1 > l2 ? nums2 : nums1;
		vector<int> &lg = l1 > l2 ? nums1 : nums2;
		if (l1 < l2) swap(l1, l2); // l1 = lg.size(), l2 = sh.size()
		if (l2 == 0) return (lg[(l1 - 1) / 2] + lg[l1 / 2]) / 2.0;

		int l = 0, r = l2 * 2;
		while (l <= r)
		{
			int mid2 = (l + r) >> 1;
			int mid1 = l1 + l2 - mid2;
			int sh_lo = (mid2 == 0) ? INT_MIN : sh[(mid2 - 1) / 2];
			int lg_lo = (mid1 == 0) ? INT_MIN : lg[(mid1 - 1) / 2];
			int sh_hi = (mid2 == l2 * 2) ? INT_MAX : sh[mid2 / 2];
			int lg_hi = (mid1 == l1 * 2) ? INT_MAX : lg[mid1 / 2];
			if (lg_lo > sh_hi)
				l = mid2 + 1;
			else if (sh_lo > lg_hi)
				r = mid2 - 1;
			else return (max(sh_lo, lg_lo) + min(sh_hi, lg_hi)) / 2.0;
		}
		return -1;
	}
};

void test()
{

}

int main()
{

	return 0;
}

File: p001-p099/leetcode004.py
class Solution(object):
	def kth(self, nums1, nums2, idx1, idx2, k):
		"""
		:type nums1: List[int]
		:type nums2: List[int]
		:type idx1: int
		:type idx2: int
		:type k: int
		:rtype: int
		"""
		l1, l2 = len(nums1), len(nums2)
		if idx1 >= l1: return nums2[k - 1]
		if idx2 >= l2: return nums1[k - 1]
		if k == 1: return min(nums1[idx1], nums2[idx2])
		if l1 - idx1 > l2 - idx2:
			return self.kth(nums2, nums1, idx2, idx1, k)
		p1 = min(l1 - idx1, k / 2)
		p2 = k - p1
		if nums1[idx1 + p1 - 1] < nums2[idx2 + p2 - 1]:
			return self.kth(nums1, nums2, idx1 + p1, idx2, p2)
		else:
			return self.kth(nums1, nums2, idx1, idx2 + p2, p1)

	def findMedianSortedArrays(self, nums1, nums2):
		"""
		:type nums1: List[int]
		:type nums2: List[int]
		:rtype: float
		"""
		l = len(nums1) + len(nums2)
		hoge = self.kth(nums1, nums2, 0, 0, l / 2 + 1)
		if l % 2:
			return hoge
		else:
			hoge += self.kth(nums1, nums2, 0, 0, l / 2)
			return hoge / 2.0
